### PR TITLE
Get sitemap product images from image cache, if available

### PR DESCRIPTION
**Pull request by sambolek**
Product images in sitemap are currently always pulled directly from gallery attributes by querying the database, whilst on frontend, we're using cached versions of the (that can potentially be stored somewhere else).

This can present an issue if google doesn't want to index images in the sitemap, as their URL is different than the one on product pages.

### Description
In sitemap, product image URLs are now being pulled by using product helper, that can serve cached image, if available. This is in line to what we see on the frontend, on product page.

### Fixed Issues (if relevant)
1. magento/magento2#7504

### Manual testing scenarios
1. Create a product and set an image on it
2. Set up and generate a sitemap
3. Observe that product image URL in sitemap is direct URL to pub/media/catalog/product http://prntscr.com/eqw82x
4. Observe that images on the product page are being pulled from image cache folder http://prntscr.com/eqw90a
5. After the fix, images are being pulled via product image helper and are in line to the way product page behaves http://prntscr.com/eqw8kl

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by adragus-inviqa** [comment]
Please use the `::class` accessor, never strings. 

E.g.
```php
// like this
ObjectManager::getInstance()->get(\Magento\Catalog\Model\Product::class);

// NOT like this
ObjectManager::getInstance()->get('Magento\Catalog\Model\Product');
```

**Comment by sambolek** [comment]
Hi @adragus-inviqa 
thanks for the suggestion, wasn't aware of this.

Got any suggestions for PHPCS reporting warnings for too long line length, like https://github.com/magento/magento2/pull/9082/commits/9c4491779cc538a633154f28a2a35540463bbe54#diff-2cc96b8b46600260b67449219e565125R76

should those lines be ignored with "@codingStandardsIgnoreStart" comments or is there another solution I missed?

**Comment by adragus-inviqa** [comment]
They all seem to have the `'http://localhost/pub/media/catalog/product/cache/c9e0b0ef589f3508e5ba515cde53c5ff/` common root, so I'd extract that into a class constant/property/anything, and concat with the rest.

For me, `@codingStandardsIgnoreStart` means `So what if it's ugly?!`: I don't particularly like it.

**Comment by sambolek** [comment]
How can I get some more info about why the previous build failed? https://travis-ci.org/magento/magento2/builds/217516099

The message doesn't help much:
`The command "test $TEST_SUITE = "static" && TEST_FILTER='--filter "Magento\Test\Php\LiveCodeTest"' || true" exited with 0.
`

**Comment by ishakhsuvarov** [comment]
@sambolek I would suggest running unit tests on local env to see what is the problem in this case.

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 1]
```diff
@@ -396,9 +396,28 @@ protected function _getAllProductImages($product, $storeId)
...
         return $this->_mediaConfig;
     }
+
+    /**
+     * Get product image URL from image filename and path
+     *
+     * @param string $image
+     * @return mixed
+     */
+    protected function getProductImageUrl($image)
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php]: I would suggest making this private, as Magento discourages inheritance-based APIs.

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 2]
```diff
@@ -396,9 +396,28 @@ protected function _getAllProductImages($product, $storeId)
      * Get media config
      *
      * @return \Magento\Catalog\Model\Product\Media\Config
+     * @deprecated No longer used, as we're getting full image URL from getProductImageUrl method
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php:421]: Please add `@see` annotation with the reference to the updated implementation.

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 3]
```diff
@@ -396,9 +396,28 @@ protected function _getAllProductImages($product, $storeId)
...
      */
     protected function _getMediaConfig()
     {
         return $this->_mediaConfig;
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php]: Looks like you have eliminated all of the usages for this dependency.
Now you could also deprecate the property and remove all type hints to it.

File: app/code/Magento/Sitemap/Model/Sitemap.php [hunk 4]
```diff
@@ -607,11 +607,12 @@ protected function _getUrl($url, $type = \Magento\Framework\UrlInterface::URL_TY
      * Get media url
      *
      * @param string $url
-     * @return string
+     * @return string $url
+     * @deprecated No longer used, as we're generating product image URLs inside collection instead
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sitemap/Model/Sitemap.php]: Please add a `@see` annotation with the reference to the new approach

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 5]
```diff
@@ -396,9 +396,28 @@ protected function _getAllProductImages($product, $storeId)
...
+
+    /**
+     * Get product image URL from image filename and path
+     *
+     * @param string $image
+     * @return mixed
+     */
+    protected function getProductImageUrl($image)
+    {
+        $productObject = ObjectManager::getInstance()->get(\Magento\Catalog\Model\Product::class);
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php]: Consider adding these as an optional constructor injections

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by sambolek** [comment]
Fixed issues mentioned in review. Waiting for CI build to finish, but tested ok locally.

Please let me know if there's anything else to fix (except potentially failed tests).

**Pull request merged**